### DISCUSSION TITLE
Other communication protocol. Air-to-Water pump (TU2C-Link?)

### DISCUSSION THREAD

**Discussion by vakkeli** [issalig/toshiba_air_cond#40] [Q&A]
Hi all, and first of all a big thanks to @issalig for the work that you have done! I will definetly buy you a beer or coffee(s)!

I have built the HW you designed and I am now able to read my Air-to-Water heat pump AB communication. It is totally different device than yours but the AB protocol seems to work with the HW you designed. However, the communication structuring seems to differ very much from your observations. I have not spent much time with this and would like to ask if anybody else has decrypted the communication that is possibly  TU2C-Link.

So if anybody else has seen something similar, please comment here. I will post updates to others here when I have had time to look at it. Perhaps finally, I will push my ready "product" in my own fork, with ESPHome and Home Assistant integrations.

Info on my project:
 - Heat pump model: **Toshiba Estia S5 HWS-805**
 - Controller model: **HWS-AMS54**
 - HW from this repo, with Wemos D1 mini
 - ESPHome with UART debug mode, baudrate 2400, parity EVEN

Some communication, power on, Heating 45C, Hot water off:

```
[11:07:28]: F0:0F:C1:E0:50:28:00:48:D5:FF
[11:07:28]: F0:10:81:E0:0C:08:00:0A:9C:FF
[11:07:30]: F0:0A:FD:FF:3A:36:FF       ^^- Typically ends message with FF
[11:07:31]: F0:0B:C1:E0:58:02:4D:FF
[11:07:31]: F0:45:FD:E0:8B:C1:13:34:99:71:2D:0B:00:07
                            ^ bit 1: HEAT, bit 2: HOT WATER maybe
[11:07:36]: F0:0A:FD:FF:3A:36:FF
...Every 5secs ^^- somehow dependent on the message size
[11:07:57]: F0:0A:FD:FF:3A:36:FF
[11:07:58]: F0:0F:C1:E0:50:28:00:48:D5:FF  Ping?
[11:07:58]: F0:10:81:E0:0C:08:00:0A:9C:FF  Pong?
[11:08:01]: F0:0B:C1:E0:58:02:4D:FF
[11:08:01]: F0:45:FD:E0:8B:C1:13:34:99:71:2D:0B:00:07
            ^^- Message start, always 0xF0
```

**Comment by issalig** [comment]
Great. I would like to integrate It in HA but i never find the time. It would be nice if you can do it. If you have any question about my code, just ask.

Maybe F0:0A ... Is a kind of ACK or status.

**Reply by issalig** [reply]
If it is only for reading values we can ignore CRC for now. 
If writing is needed then I suggest you to play with https://www.scadacore.com/tools/programming-calculators/online-checksum-calculator/

**Comment by vakkeli** [comment]
Short update, will go in-depth later:
 - I whipped out my oscilloscope and loaned a Windows laptop (no linux sw with the scope :( )
 - Turns out there is no parity bit, one stop bit and baud rate confirmed to be 2400.
 - I am able to read if the device is on/off, heating water on/off, heating hot-water on/off, and is it using heat pump and/or resitor to do so. Target temperatures.
 - First bytes always [0xF0,  0xF0] leading with length and last byte always 0xA0, Checksum is CheckSum8 modulo 256 which is the byte before 0xA0

Status message example:
F0:F0:15:70:FF:E0:A0:31:C0:00:00:A2:70:66:68:52:58:00:00:7F:A0

The temperatures follow specific function: (value / 2 - 16) celsius. I have not figured out why, as bitwise this does not make sense?

Now I need to debug the write logic on HW level, because I think I am not able to write anything there. @issalig any tips for debugging the write logic. I assume that the writing can be also observed from the read side optocoupler leg and ground?

**Reply by vakkeli** [reply]
Got me thinking about the temperature values, that it probably has something to do with Farenheit compability and the values are sent as Farenheit as the conversion formula is: F=C*1.8+32

**Reply by Etacovda** [reply]
I was going to say, its probably to do with farenheit :)

I'm literally going down this same path now to do what you're doing - with the same heatpump (well, the 14kw version). Watching with much interest. 

How did you build the hardware, did you get a prototype pcb manufactured, or just solder up on some perf or similar?
This is a new adventure for me, I haven't had boards manufactured before. Looking forward to getting it into home assistant or node-red.

**Reply by issalig** [reply]
Interesting, in my system temp is value/2 -35 https://github.com/issalig/toshiba_air_cond/blob/6a1a4507f1c999631f7ad3b899e45bc7a3127b4c/air/toshiba_serial.hpp#L304

In order to send a message, just listen to a message when for example you change the temperature and send those bytes.

For example for sending the message you identified as a PING (F0:0F:C1:E0:50:28:00:48:D5:FF) you can write a function like this one (and put it at the end of toshiba_serial.hpp, and call it from the main loop and see it it receives pong):

void air_send_test_ping(air_status_t *air) {
  int i;

  SoftwareSerial *ss;
  ss = &(air->serial);

  const unsigned char testdata[] = {
    0xF0,0x0F,0xC1,0xE0,0x50,0x28,0x00,0x48,0xD5,0xFF
  };

  ss->enableIntTx(true);

  Serial.println("");
  Serial.print("Sending test data (");
  Serial.print(sizeof(testdata));
  Serial.println(")");

  for (i = 0; i < sizeof(testdata); i++) {
    ss->write(testdata[i]);
    //    Serial.print(testdata[i] < 0x10 ? " 0" : " ");
    //    Serial.print(testdata[i], HEX);
  }
  Serial.println("");

  ss->enableIntTx(false);

}

**Reply by vakkeli** [reply]
@Etacovda I ordered a batch of PCBs from JLCPCB.

**Reply by vakkeli** [reply]
@issalig On the software side I have things in order (I guess). I don't get any response from the pump, eventhough I have tried multiple message types. I am starting to think that there is something wrong with the HW writing side. I need to scope the writing and see if it goes through.

I am also experimenting to add this DIY controller as secondary remote, as that is an option at least in our series of pumps. It does some kind of pairing routine when switching the Header remote to Secondary. I need to go through this communication and see if I can mimic that. Taken that I get the writing to work.

As a backup plan I had previously installed some wires to button pads to the Header remote, so I can use some optocouplers to "press" those buttons.

**Reply by issalig** [reply]
Yes, JLCPCB has good prices and it takes around or less than 2 weeks to arrive at your place (which I think is great).

Moreover, I still have some spare PCB's and components if you are interested.

**Reply by issalig** [reply]
"As a backup plan I had previously installed some wires to button pads to the Header remote, so I can use some optocouplers to "press" those buttons."

Hehe, yes, there is always a backup plan.

**Comment by vakkeli** [comment]
Again quick update:

Write is now working! I am able to turn the device on and off with static string commands.

My optocoupler and/or Q2 on the writing side was fried. I am not sure which one was it in the end. At least multimeter shows dead LED in the coupler.

CAUTION to everyone! Check first which one is positive and which one is negative. In my case A was actually negative and B was positive. In this repos pictures A connects to the AB-link positive side.

**Reply by issalig** [reply]
I have added a warning in the instructions.

**Comment by Etacovda** [comment]
Yep, I would be keen - seems silly to get 5 of them if you have spares! Let
me know how to pay you and get an address to you and I’ll definitely get
one

**Comment by vakkeli** [comment]
I as well have two sets of PCBs and components I can send. Excluding the ESP8266, Diodes and resistors tho.
You can send me a telegram message

**Comment by Etacovda** [comment]
hey vakkeli, that would be great with components etc, makes it simple - how
would you like to organise it? happy to paypal or something like that?

thanks,

**Reply by vakkeli** [reply]
I don't like to step on @issalig toes here, if he has PCBs and components laying around, maybe he can send it to you. If not, I am happy to help. You can telegram me.

**Reply by issalig** [reply]
Don't worry @vakkeli , it is nice to see user collaboration.
If any of you guys, have experience with HA I would like to make my software work there. So any help will be appreciated.

**Reply by vakkeli** [reply]
After I am finished I will turn my repo/fork public and you can copy it from there. As said I am using ESPHome and I have already built a custom component. I have not actually used any SW code from this repo, built from ground up. Custom component is built around C++ and some Python and your code can be reused there. ESPHome makes the integration to HA basically effortless, it just shows up there. ESPHome needs to be ran separetly tho in a Docker or similar.

Current status: I have all the readable information passed to HA sensor, now I am working on getting the Switches (On/off, modes) and Inputs (temperatures) to work and polling for the additional data that can found from the field menu.

@Etacovda and others can send me a telegram message if they are intrested in buying the PCB.

**Comment by Etacovda** [comment]
I'm happy to make a node-red flow to integrate it into mqtt or similar as
well, if that's helpful (ie, direct node red, without esphome - but if theres an esphome version ill probably use that
as the majority of my house is run off esphome!) I'll probably order a POE powered ethernet esp32, seems to be a nice way to
do it; my heating systems getting pretty extreme now in its complexity...!

I wonder, can you enable the automatic mode from an address? thats one of my major annoyances with this heatpump, is that you can schedule fixed temperature heating on and off, but not the weather compensated version...

thanks!

**Reply by vakkeli** [reply]
Yes, I have not yet tested it but I have observed the command to turn Auto mode on and off. You can also control the temperature directly from HA or similar to make more "Intelligent" Auto mode yourself.

For PoE boards I highly recommend [Olimex boards](https://www.olimex.com/Products/IoT/ESP32/ESP32-POE/open-source-hardware). Relatively cheap and not those Ali knock-offs! You can also use them without poe wired and even with battery (charger onboard). The pins might even be compatible with this PCB. Not sure though, need to check first!

**Reply by Etacovda** [reply]
Ah, thats great having all those options.

I figure I'm going to make an automatic mode anyway for cooling - I have a two story house with underfloor in concrete slab downstairs, and using aluminium spreader plates upstairs, as well as fan coil units upstairs (living room etc is upstairs, most bedrooms downstairs). This integration will allow me to make calculations on dew point and run the water temperature at just above that, and cool the house safely without fear of condensation, which will be very nice.

That esp board does look very nice (and like you said, not a bad price either!)

@issalig  - do you have a paypal address or something that I can send you some money to for your work? I'm not that familiar with github, but I can't see any way to send you some money unless its not obvious...!

**Reply by issalig** [reply]
@Etacovda , thank you, I do have paypal and I appreciate your offer. If you can help this project to make it better this is the best payment.

**Comment by Etacovda** [comment]
> @Etacovda , thank you, I do have paypal and I appreciate your offer. If you can help this project to make it better this is the best payment.

sounds good - I will share whatever I do along with photos and any code. I'll probably hijack off of vakkeli's code for the majority of it and fork that.

**Comment by vakkeli** [comment]
I have pushed my unfinished code to my own ESPHome repo: https://github.com/vakkeli/toshiba_uart_ctrl

Few people have been intrested, so there you go. A disclamer though, It's still very much unfinished. Fortunately the reason I have not finished it, is that it has been running with my home automation and HA perfectly for four months now, wihtout any issues. So even though it's not finished it is very usable. :D

Only works for the Estia Air-to-Water pumps to my knowledge.

**Comment by Etacovda** [comment]
FYI finally got around to finishing mine (depths of 'winter' here), and its working with your code vakkeli, thanks!
